### ISSUE TITLE
LearningRateFinder during training

### ISSUE THREAD

**Issue by X-Bruce-Y** [Lightning-AI/pytorch-lightning#20008]
### Description & Motivation

LearningRateFinder callback is cool to find a learning rate that is potential to reduce training loss from the start. However, in my case since epoch 2, learning rates need to decreased a lot promptly, otherwise the loss would not drop any more. That makes me wonder why the technique is not extended to the whole process of training. Though that would take longer (presumbaly < 2x) for each epoch, the overall efficacy should be expected to improve.

### Pitch

Extend the LearningRateFinder callback so that it can be called during the whole training process, with user-defined interval (probably using epoch as unit), start_epoch, end_epoch, etc.

### Alternatives

Currently it's possible to train one epoch, save the checkpoint, re-initiate the model with the latest epoch, use the callback to find optimal learning rate, and do it all over again. However, it's so indirect and slows training down too much.

cc @lantiga @borda

**Comment by SkafteNicki** [comment]
Thanks for opening this issue. I think it is unclear how beneficial to the overall training efficacy it would be to re-find the learning rate during training, but that is definitely something that could be investigated. Since we cannot really predict at what other points during training that users may want to call the learning rate scheduler, I am convinced that the better option here is to use the customization of the callback system to implement custom learning rate finders based on what behavior you want. For example, this `MileStoneLearningRateFinder` runs on epoch start for user specified epoch numbers:
```python
class MilestoneLearningRateFinder(LearningRateFinder):
    """
    LearningRateFinder that runs at specific milestone epochs.
    
    Usage:
        lr_finder = MilestoneLearningRateFinder(milestone_epochs=[0, 5, 10])
        trainer = Trainer(callbacks=[lr_finder])
    """
    
    def __init__(self, milestone_epochs: List[int], **kwargs):
        super().__init__(**kwargs)
        self.milestone_epochs = set(milestone_epochs)
    
    def on_fit_start(self, trainer: "pl.Trainer", pl_module: "pl.LightningModule") -> None:
        if 0 in self.milestone_epochs:
            super().on_fit_start(trainer, pl_module)
    
    def on_train_epoch_start(self, trainer: "pl.Trainer", pl_module: "pl.LightningModule") -> None:
        if trainer.current_epoch in self.milestone_epochs and trainer.current_epoch > 0:
            self.lr_find(trainer, pl_module)
```
and similar changes can be made for other custom patterns.
I am therefore going to close the issue, but ping me and reopen if more uses are requesting that this feature gets build directly into the learning rate finder in lightning.
